stumpy/stumped.py

```diff
@@ -248,20 +253,41 @@ def stumped(dask_client, T_A, m, T_B=None, ignore_trivial=True, normalize=True,
                 T_B_subseq_isconstant_future,
                 diags_futures[i],
                 ignore_trivial,
+                k,
             )
         )
 
     results = dask_client.gather(futures)
-    profile, indices = results[0]
+    profile, indices, profile_L, indices_L, profile_R, indices_R = results[0]
+
     for i in range(1, len(hosts)):
-        P, I = results[i]
-        for col in range(P.shape[1]):  # pragma: no cover
-            cond = P[:, col] < profile[:, col]
-            profile[:, col] = np.where(cond, P[:, col], profile[:, col])
-            indices[:, col] = np.where(cond, I[:, col], indices[:, col])
-
-    out[:, 0] = profile[:, 0]
-    out[:, 1:4] = indices
+        P, I, PL, IL, PR, IR = results[i]
+        for j in range(l):
```

Comment:
I don't know. This set of nested for-loops and if statements seem too complicated. We should try to replace them with `numpy` operations where possible so that it may be easier to follow/process

Reply:
@seanlaw 
I had an idea but I think it requires additional memory and maybe it is more time consuming. I was thinking of something like:
```
IDX = np.argsort(np.concatenate((profile, P), axis=1), axis=1, kind='mergesort')[:k]
#then update profile and I with help of IDX 
```

However, this needs new memory to store `np.concatenate(...)` and the sorting is going to sort `2*k` elements which may be redundant in some cases.

---

I will think more about it to see if I can come up with a clean yet efficient approach.

Reply:
How do you feel about this? (%100 pass in `test_stumped.py` )

```
offset = np.arange(k, dtype=np.int64)
mask = np.full(k, 1, dtype=bool)
for i in range(1, len(hosts)):
    P, I, PL, IL, PR, IR = results[i]
    for j in range(l):
        mask[:] = True
        IDX = np.searchsorted(profile[j], P[j], side="right")
        IDX = IDX + offset
        IDX = IDX[IDX < k]
        mask[IDX] = False

        profile[j, mask] = profile[j, : k - len(IDX)].copy()
        profile[j, ~mask] = P[j, : len(IDX)]

        indices[j, mask] = indices[j, : k - len(IDX)].copy()
        indices[j, ~mask] = I[j, : len(IDX)]
```


Is there any way to avoid `.copy()` in code above? If I do not use `.copy()`, it shows weird behavior:
```
# example
mask = np.array([False,  True, False,  False,  True])
profile = np.array([0,1,2,3,4])

profile[mask] = profile[:2]
profile
>>> array([0, 0, 2, 3, 0])

# expected result: array([0, 0, 2, 3, 1])
```

Reply:
Yeah, it's likely because the writing to `profile[mask]` is happening sequentially and not simultaneously. So `profile[1]` gets set to zero first and then `profile[4]` gets set to `profile[1]`, which has already been set to zero.

Reply:
> How do you feel about this?

I'd like to think about this a little more

Reply:
Just in case that matters, I am sharing the computational burden of `.copy()` below:
```
import time
import numpy as np
A = np.random.rand(50) # matrix profile
B = np.random.rand(50) # matrix profile indices
n_itr = 1000000

tic = time.time()
for i in range(n_itr):
    A_copy = A.copy()
    B_copy = B.copy()
toc = time.time()
toc - tic
>>> 0.93
```

Reply:
@NimaSarajpoor Can we do something simpler like this:

```
import numpy as np

n = 3
k = 5

results = [
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k)),
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k)),
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k))
]

profile = np.empty((n, 2 * k))
indices = np.empty((n, 2 * k), dtype=np.int64)
profile[:, :k], indices[:, :k] = results[0]

for i in range(1, len(results)):
    P, I = results[i]
    profile[:, k:] = P
    idx = np.argsort(profile, axis=1)
    profile[:, :] = np.take_along_axis(profile, idx, axis=1)
    indices[:, :] = np.take_along_axis(indices, idx, axis=1)
    
profile = profile[:, :k]
indices = indices[:, :k]
```

Reply:
Btw, I'm not saying that this is faster but I find this to be much more readable

Reply:
I really enjoy how you see things!!!

> Btw, I'm not saying that this is faster but I find this to be much more readable

I see your point! Cool!

---

If I may ask, is there any particular reason we do `profile[:, :] = np.take_along_axis(...)` instead of `profile = np.take_along_axis(...)`?

Reply:
I don't want to make it seem like I just know how to do this stuff. You should know that there was a lot of Google searching and trial-and-error before I was satisfied with this solution.

> If I may ask, is there any particular reason we do profile[:, :] = np.take_along_axis(...) instead of profile = np.take_along_axis(...)? Is it okay to go with the latter?

No! The former, `profile[:, :]` copies the results over and into the existing profile memory space. I suspect that the latter creates a brand new array in memory and then assigns the variable name `profile` to that new array.

Reply:
> I don't want to make it seem like I just know how to do this stuff. You should know that there was a lot of Google searching and trial-and-error before I was satisfied with this solution. I have learned a lot from other StackOverflow experts!


Still, you are very good at connecting dots in your mind!



> No! The former, profile[:, :] copies the results over and into the existing profile memory space.


Right!!! I overlooked that! Thanks for your thorough explanation.

Reply:
Here is another (different) option that is parallelized with `numba`:

```
from numba import njit, prange
import numpy as np

@njit(parallel=True)
def merge_topk_profiles_indices(PA, PB, IA, IB):
    for i in prange(PA.shape[0]):
        for j in range(PA.shape[1]):
            if PB[i, j] < PA[i, -1]:
                idx = np.searchsorted(PA[i], PB[i, j], side="right")
                if idx < PA.shape[0]:
                    PA[i, idx + 1 :] = PA[i, idx:-1]
                    PA[i, idx] = PB[i, j]
                    IA[i, idx + 1 :] = IA[i, idx:-1]
                    IA[i, idx] = idx

                    
n = 3
k = 5

results = [
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k)),
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k)),
    (np.sort(np.random.rand(n*k).reshape(n, k), axis=1), np.arange(n*k).reshape(n, k))
]

profile, indices = results[0][0], results[0][1]

for i in range(1, len(results)):
    P, I = results[i]
    merge_topk(profile, P, indices, I)
```

It will look very similar to what you had before but I've abstracted it out into a function that should be much easier to follow/interpret/read

Reply:
> @njit(parallel=True)
def merge_topk_profiles_indices(PA, PB, IA, IB)

Cool! Yes, the name of this function alone helps to understand what is going on! Btw, I think we can remove `if idx < PA.shape[1]:` as the condition is always True.

What do you think? Should I try them out and see which one performs better on `l=1000000`  and `k=50`?

Reply:
Does the `numba` one give the right answer? If so, then I think it would make sense to go with `numba`

> Btw, I think we can remove if idx < PA.shape[1]: as the condition is always True.

Really? What happens when the value in `PB` is larger than (or equal to) all values in `PA`? Doesn't that make `idx == PA.shape[0]`?

Reply:
> Does the numba one give the right answer? If so, then I think it would make sense to go with numba.

Sounds good. I will try it out and see how it goes.

> Really? What happens when the value in PB is larger than (or equal to) all values in PA? 

I think that's not going to happen as it will be caught by the outer if, which is `if PB[i, j] < PA[i, -1]:`. So, the maximum value for `idx` is `PA.shape[0] - 1`.

Reply:
You're right. Thank you

Reply:
I think the `argsort` version could be the slower "naive" version

Reply:
@seanlaw 
Yes....the `argosrt` version is  slower. Just to make sure I am on the same page as you are... we are not talking about creating naive version for `gpu_stump` though. Right? The test function of `gpu_stump` uses `naive.stump`.

Reply:
Correct. I meant a `naive.merge_topk_PI` using `argsort` and then the performant version is using `np.searchsorted`